src/lib/shell/directory_stack.rs

```diff
@@ -235,18 +235,35 @@ impl DirectoryStack {
         dir: &str,
         variables: &Variables,
     ) -> Result<(), Cow<'static, str>> {
-        match (set_current_dir(dir), current_dir()) {
-            (Ok(()), Ok(cur_dir)) => {
-                self.push_dir(cur_dir, variables);
+        use std::path::{Component, Path};
+
+        // Create a clone of the current directory.
+        let mut new_dir = match self.dirs.front() {
+            Some(cur_dir) => cur_dir.clone(),
+            None => PathBuf::new()
+        };
```

Comment:
Note that you can write this as:

```rust
let new_dir = self.dirs.front().map_or(PathBuf::new(), |dir| dir.clone());
```

Reply:
Won't using map_or cause PathBuf::new() to be evaluated eagerly so that it gets instantiated every time regardless of whether there is a current directory or not?

Reply:
Yeah, `map_or_else` is generally what Clippy will bark at you to use.

Reply:
Which reminds me that Clippy needs to be run on Ion's code base, and all issues reported by Clippy addressed.